copy/copy.go

```diff
@@ -309,6 +323,18 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
 	// If src.UpdatedImageNeedsLayerDiffIDs(ic.manifestUpdates) will be true, it needs to be true by the time we get here.
 	ic.diffIDsAreNeeded = src.UpdatedImageNeedsLayerDiffIDs(*ic.manifestUpdates)
 
+	// Unblock all blobs as a precation to avoid potential deadlocks in the
+	// presence of errors or even panics.  Note that it is safe to unblock
+	// here as locks are only unblocked if we're the owner of the lock;
+	// only the first call to UnlockBlob() will effectively unlock the file.
+	if ic.c.dest.SupportsBlobLocks() {
+		defer func() {
+			srcInfos := ic.src.LayerInfos()
+			for _, blob := range srcInfos {
+				ic.c.dest.UnlockBlob(blob)
```

Comment:
Why isn’t every layer unlocked as soon as that one blob is copied? 

(… because only `storageImageDestination.Commit` actually moves creates a c/storage layer from the temporary file, if I’m not mistaken. Maybe that should be changed first? Maybe not. At the very least, it should be explicitly and clearly documented why there isn’t a single function / level of the call stack both locking and unlocking.)

Reply:
Exactly. Committing the layer happens at a different part of the stack and I am more than afraid of the complexity changing that. It's definitely possible to change the way we commit but this would imply rewriting a huge amount of code.

I had the very same thought when thinking about that and from an architectural point of view I'd prefer to have those tasks tied together but I currently don't have the time to refactor that.

Reply:
I will update the comments accordingly and be verbose about that. Thanks for bringing that up!

Reply:
This has been changed as remaining locks are now unlocked in `dest.Close()` while the storage destination keeps track of which blobs/digests have actually been locked.